src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +1003,762 @@ struct
   let to_yojson (x, y) = `Assoc [ (Base.name (), Base.to_yojson x); ("length", Idx.to_yojson y) ]
 end
 
-module AttributeConfiguredArrayDomain(Val: LatticeWithSmartOps) (Idx:IntDomain.Z):S with type value = Val.t and type idx = Idx.t =
+module HelperFunctionsIndexMustMaySets =
+struct
+  module MustSet = SetDomain.Reverse (SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All indexes" end))
+  module MaySet = SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All indexes" end)
+
+  let compute_set len =
+    List.init (Z.to_int len) (Fun.id)
+    |> List.map Z.of_int
+    |> MustSet.of_list
+
+  let must_nulls_remove i must_nulls_set min_size =
+    if MustSet.is_bot must_nulls_set then
+      MustSet.remove i (compute_set min_size)
+    else
+      MustSet.remove i must_nulls_set
+
+  let must_nulls_filter cond must_nulls_set min_size =
+    if MustSet.is_bot must_nulls_set then
+      MustSet.filter cond (compute_set min_size)
+    else
+      MustSet.filter cond must_nulls_set
+
+  let must_nulls_min_elt must_nulls_set =
+    if MustSet.is_bot must_nulls_set then
+      Z.zero
+    else
+      MustSet.min_elt must_nulls_set
+
+  let may_nulls_remove i may_nulls_set max_size =
+    if MaySet.is_top may_nulls_set then
+      MaySet.remove i (compute_set max_size)
+    else
+      MaySet.remove i may_nulls_set
+
+  let may_nulls_filter cond may_nulls_set max_size =
+    if MaySet.is_top may_nulls_set then
+      MaySet.filter cond (compute_set max_size)
+    else
+      MaySet.filter cond may_nulls_set
+
+  let may_nulls_min_elt may_nulls_set =
+    if MaySet.is_top may_nulls_set then
+      Z.zero
+    else
+      MaySet.min_elt may_nulls_set
```

Comment:
Why not put them inside your may/must modules respectively?

~~~ocaml

module MustSet = struct 
  include SetDomain.Reverse (SetDomain.ToppedSet (IntDomain.BigInt) (struct let topname = "All indexes" end))
  (* the functions for MustSet *)
  
  let filter cond must_nulls_set min_size =
    if MustSet.is_bot must_nulls_set then
      MustSet.filter cond (compute_set min_size)
    else
      MustSet.filter cond must_nulls_set
end
~~~

Then you can directly call things like `MustSet.filter` which should make the code nicer to read.